x/uniswap/handler.go

```diff
@@ -0,0 +1,216 @@
+package uniswap
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/supply/types"
+)
+
+// NewHandler routes the messages to the handlers
+func NewHandler(k Keeper) sdk.Handler {
+	return func(ctx sdk.Context, msg sdk.Msg) sdk.Result {
+		switch msg := msg.(type) {
+		case MsgSwapOrder:
+			return HandleMsgSwapOrder(ctx, msg, k)
+		case MsgAddLiquidity:
+			return HandleMsgAddLiquidity(ctx, msg, k)
+		case MsgRemoveLiquidity:
+			return HandleMsgRemoveLiquidity(ctx, msg, k)
+		default:
+			errMsg := fmt.Sprintf("unrecognized uniswap message type: %T", msg)
+			return sdk.ErrUnknownRequest(errMsg).Result()
+		}
+	}
+}
+
+// HandleMsgSwapOrder handler for MsgSwapOrder
+func HandleMsgSwapOrder(ctx sdk.Context, msg MsgSwapOrder, k Keeper) sdk.Result {
+	var caclulatedAmount sdk.Int
+
+	// check that deadline has not passed
+	if msg.ctx.BlockTime.After(msg.Deadline) {
+		return ErrInvalidDeadline(DefaultCodespace, "deadline has passed for MsgSwapOrder")
+	}
+
+	if msg.IsBuyOrder {
+		calculatedAmount := k.GetInputAmount(ctx, msg.SwapDenom, msg.Amount)
+		// ensure the maximum amount sender is willing to sell is less than
+		// the calculated amount
+		if msg.Input.Amount.LT(calculatedAmount) {
+			return types.ErrInvalidBound(DefaultCodespace, "maximum amount (%d) to be sold was exceeded (%d)", msg.Input.Amount, calculatedAmount).Result()
+		}
+
+		coinSold := sdk.NewCoins(sdk.NewCoin(msg.Input.Denom, calculatedAmount))
+		if !s.bk.HasCoins(ctx, msg.Sender, coinSold) {
+			return ErrInsufficientAmount(DefaultCodespace, "sender account does not have sufficient funds to fulfill the swap order").Result()
+		}
+
+		err := k.sk.SendCoinsFromAccountToModule(ctx, msg.Sender, ModuleName, coinSold)
+		if err != nil {
+			return err.Result()
+		}
+
+		err = k.sk.SendCoinsFromModuleToAccount(ctx, ModuleName, msg.Sender, sdk.NewCoins(msg.Output))
+		if err != nil {
+			return err.Result()
+		}
+
+	} else {
+		outputAmount := k.GetOutputAmount(ctx, msg.SwapDenom, msg.Amount)
+		// ensure the minimum amount the sender is willing to buy is greater than
+		// the calculated amount
+		if msg.Output.Amount.GT(outputAmount) {
+			return sdk.ErrInvalidBound(DefaultCodespace, "minimum amount (%d) to be sold was not met (%d)", msg.Output.Amount, calculatedAmount).Result()
+		}
+
+		coinSold := sdk.NewCoins(msg.Input)
+		if !s.bk.HasCoins(ctx, msg.Sender, coinSold) {
+			return ErrInsifficientAmount(DefaultCodespace, "sender account does not have sufficient funds to fulfill the swap order").Result()
+		}
+
+		err := k.sk.SendCoinsFromAccountToModule(ctx, msg.Sender, ModuleName, sdk.NewCoins(msg.Input))
+		if err != nil {
+			return err.Result()
+		}
+
+		err = k.sk.SendCoinsFromModuleToAccount(ctx, ModuleName, msg.Sender, sdk.NewCoins(sdk.NewCoin(msg.Output.Denom, calculatedAmount)))
+		if err != nil {
+			return err.Result()
+		}
+
+	}
+
+	return sdk.Result{}
+}
+
```

Comment:
I was going through 'GetInputPrice` and `GetOutputPrice` mentioned in https://github.com/runtimeverification/verified-smart-contracts/blob/uniswap/uniswap/x-y-k.pdf

After discussing with @AdityaSripal we both feel that you swapped `lesser than` and `greater than` when it comes to buy order and sell order.

For instance, for a buy order, the calculated amount should be lesser than the maximum amount the sender is willing to pay. Thus change `if msg.Input.Amount.LT(calculatedAmount)` to `if calculatedAmount.LT(msg.input.Amount)` 

Similarly for a sell order, the calculated amount should be greater than the minimum amount the sender is willing to pay. Thus change `if msg.Output.Amount.GT(outputAmount)` to `if calculatedAmount.GT(msg.Output.Amount)`. I would change `outputAmount` to `calculatedAmount` for uniformity.

A better way to write the comment would be to specify calculated amount first followed by the amount the sender is willing to pay.